drivers/fuel_gauge/max17048/max17048_trigger.c

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (c) 2023 Fraunhofer AICOS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/drivers/fuel_gauge/max17048.h>
+#include "max17048.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(MAX17048);
+
+/* Forward declaration of internal functions */
+static int max17048_read_register(const struct device *dev, uint8_t registerId, uint16_t *response);
+static int max17048_update_register(const struct device *dev, uint8_t reg, uint16_t mask,
+				    uint16_t val);
+
+static int max17048_clear_alert(const struct device *dev);
+static void max17048_int_callback(const struct device *port, struct gpio_callback *cb,
+				  uint32_t pin);
+static void max17048_int_work(struct k_work *work);
+
+static int max17048_undervoltage_threshold_set(const struct device *dev, const uint16_t voltage);
+static int max17048_overvoltage_threshold_set(const struct device *dev, const uint16_t voltage);
+static int max17048_low_soc_threshold_set(const struct device *dev, const uint8_t soc);
+
+/* Public interface */
+int max17048_trigger_set(const struct device *dev, const enum max17048_trigger_type trigger_type,
```

Comment:
Just a bit of added context: It seems to me that normally we don't deal with userspace added callbacks. However, I don't want to preclude that from the fuel gauge API for now.

Reply:
Callbacks indeed are not available to userspace, historically the way other APIs deal with this is using k_poll_signal.

The I2C API (and in looking this up I noticed the docs are broken, will be fixing...) offers both a k_poll_signal and callback style API.

The way sensors allows for callbacks is by pushing the callback off to a work queue task or dedicated sensor thread. This has some undesirable implications (e.g. latency or memory usage trade offs).

We could implement tock's "upcall" idea maybe one day... but frankly I'm not a big fan of that idea as it would result in a lot of added context swaps the way I picture it working.